a11y: Improve accessibility of dropdown menus

Jookia opened the pull request:
This patch set fixes keyboard and screen reader accessibility for dropdown menus and lays the groundwork for improvements to other dropdown-related components such as list boxes and search menus.

Note that this doesn't change any behaviour of the dropdown menus themselves outside focusing, but the dropdown menus do seem a little strange compared to dropdown menus on other sites as they automatically open and retain their selected item between focuses.

To test this patch, just verify that:
- Dropdowns still work
- Only dropdown menus have changed HTML
- Dropdown icons next to dropdown menus are no longer focusable
- Screen readers can hear dropdown menus

guillep2k commented:
Please address the lint errors from the build: https://drone.gitea.io/go-gitea/gitea/14488/1/7

You do the checks locally using:
```
make js
```

Jookia commented:
I'll fix the lint errors that are mine, but a lot are from Semantic's dropdown.js. What should be done about that?

guillep2k commented:
> 
> 
> I'll fix the lint errors that are mine, but a lot are from Semantic's dropdown.js. What should be done about that?

That's a good question. We normally use the /vendor directory to place external libraries, which are exempt from linting. But only when they are untouched (e.g. right from npm). Is that the case here, or you had to edit it to make it work? (I can imagine the answer).

Jookia commented:
The semantic.dropdown.js is taken from Semantic UI then edited to add accessibility features, so it's definitely not from upstream- in commit #2 of this patch series I explicitly added a note at the top of the file that it's been modified for Gitea.

Jookia commented:
I've found a bug in this: aria-expanded and aria-activedescendant are set regardless of the current role, so I'll have to fix that too.

Jookia commented:
I also realized I use double quotes instead of single quotes in a lot of places, and more style fixes.

guillep2k commented:
@Jookia No problem. You can edit the title and prefix it with `WIP: ` or `[WIP] ` (with a space), but not `[WIP]: ` for some reason 😄.

Jookia commented:
Okay, I think all my issues are fixed. Linting shows no errors from my lines of code either.

zeripath commented:
Finally had time to run this.

There's an issue with using the keyboard on any of the dropdowns in that the visible label of dropdown changes including losing the down arrow. You probably need to set the aria-label instead.

Jookia commented on templates/user/dashboard/navbar.tmpl:
```diff
@@ -11,18 +11,16 @@
 					<div class="ui header">
 						{{.i18n.Tr "home.switch_dashboard_context"}}
 					</div>
-					<div class="scrolling menu items">
```
I didn't remove just the scrolling, I removed the entire sub container. How would I test to see if it's large enough to fill the screen?

zeripath replied:
Just create more organizations

Jookia commented on public/js/semantic.dropdown.js:
```diff
@@ -0,0 +1,4019 @@
+ *
+ */
+
+/* This version has been modified for Gitea */
```
Can I leave out the 'All rights reserved.'? Seems a bit weird given it's under the MIT license

zeripath replied:
I think you have to still put it as it's standard boilerplate. I'd probably put the comment before the copyright so as to make it clear that only the modifications are Gitea's but it's ok

Jookia commented:
@guillep2k This should only affect dropdown menus.
@zeripath I do set the aria-label appropriately. The problem you're noticing is present in the existing codebase and not introduced by these commits. It's strange behavior, but that's how Semantic works.

zeripath commented:
@Jookia Ugh I think that needs fixing too! If it's too hard to get sorted then maybe we can fix it later(?)

Jookia commented:
I don't think it's a fix for this PR, since it requires a lot more research of the code base.

To elaborate: Semantic doesn't understand what an actual 'menu' is, it just knows about dropdowns and some mixins like having it be a form or have the items search, etc. So when used for navigation menus, it changes the value on the actual item just as you would when selecting an element in a list box.

Sometimes it does this when you click on some dropdown menus too, such as the languages at the bottom of the screen. It doesn't do it other times when you click on things such as sort buttons or the navigation buttons. The logic behind this is something I would have to dig in to.

We could just say 'dropdowns that don't set a value (aka menus) shouldn't remember their currently selected item', which would reflect them being used for menus better- after all, if they don't set a value they're likely going to contain links or actions instead.

But sometimes you want them to remember their value when you use the menus for things like repository collaboration permissions which is a plain 'menu' but its actions don't take you away from the page so you want its selection to stay.

Jookia commented:
To be clear: I really don't want to change any user visible behavior in this PR, just make it accessible. I feel like changes to behavior would need long discussion about what the right behavior is in an issue, and I don't want that to impede this PR's goal of just having the menus accessible to keyboards and screen readers. If you want I can do up an issue explaining what UI problems dropdown menus have an open the discussion there.

zeripath commented:
No worries - I understand what you're saying. The menus changing their title when you tap through them is a problem for accessibility (but not all of our dropdowns act as menus.) However I think there's enough in the pr already - it doesn't need to fix everything!

I was just worried that was a regression but as it's not that's great.

zeripath commented:
I agree with @guilep2k comments. I'll have another play to see if I can break anything but if you address those and I can't break anything then I would approve

Jookia commented:
Ok, new version. I fixed the version note (see the last commit) and removed the dashboard menu 'fix' that broke scrolling. I also removed the patch to set tabindex on child dropdown icons as it didn't work for the dashboard. Both of these changes make accessibility a little worse, but my reasoning is that these are workarounds for specific issues I've seen- the only way I'm going to truly fix these weird dropdown menus is to manually look through the HTML templates and fix those.

Jookia commented:
I merged in the MD5 version marker suggestions, though I haven't tested them yet.

silverwind commented on Makefile:
```diff
@@ -434,7 +434,7 @@ npm-update: npm-check
 .PHONY: js
 js: npm
-	npx eslint public/js
+	npx eslint --ignore-pattern semantic.dropdown.js public/js
```
Use `.eslintignore` instead.

silverwind replied:
Or actually, move vendor lib to `public/vendor`, it has no place in `public/js`.

guillep2k replied:
@silverwind it's not "exactly" vendor, since it's been specifically modified for Gitea. It's now Gitea code.
One possibility is to move it to its own project @ gitea.com.

Jookia replied:
There's no .eslintignore file?

silverwind replied:
You can create one. I suggest renaming the file to `semantic.dropdown.custom.js` to highlight that it's a customized version and then create `.eslintignore` with content `/ public/js/semantic.dropdown.custom.js`.

Jookia replied:
Ah ok, will try that.

Jookia replied:
Ok, done!

silverwind replied:
Looking good, thanks

silverwind commented:
140kB seems a bit of a big dependency, is it really worth it?

guillep2k commented:
> 
> 
> 140kB seems a bit of a big dependency, is it really worth it?

I believe it replaces the currently used version of semantic dropdown; in theory it shouldn't produce a noticeable delta (but I didn't check).

zeripath commented:
I checked - it works.

Jookia commented:
This doesn't replace the existing semantic dropdown code, it just overrides it with a copy of unminified dropdown code. So it effectively duplicates and unminifies semantic.js' dropdown code. Since it's intended to be modified, minifying it at this point isn't an option.

I'm not sure what you mean by 'is it really worth it'- do you mean sending unminified JS and duplicating some code in the repo so disabled people being able to used gitea?

The alternative is for me to just fork semantic, but that would mean pulling in new minified JS with every PR I do to fix bugs in semantic and gitea.

guillep2k commented:
> 
> 
> This doesn't replace the existing semantic dropdown code, it just overrides it with a copy of unminified dropdown code. So it effectively duplicates and unminifies semantic.js' dropdown code. Since it's intended to be modified, minifying it at this point isn't an option.

@Jookia size of source code is not a concern; only size of code delivered to clients.

Webpack was PR'd very recently for index.js (#8440), it just needs to be reviewed. I'm sure this PR's code can be minified as well.

zeripath commented:
@silverwind there is, as far as I can see, no other way of improving a11y except by overriding the whole of drop-down - there are no hooks that we can use. We can't offer a pr upstream as upstream appears to be dead. Screen-reader users cannot use Gitea without this PR.

As a possible upstream, Fomantic-ui would likely take a pr but it isn't a drop in replacement for semantic-ui - I tried. A more CSS savvy person might be able to reason about what needs to be changed but it seemed like there were multiple problems which would necessitate close examination of every page of Gitea - meaning that a PR that should take 3 minutes could easily take days to check.

I agree we should be minifying, but Minification should be dealt with in the webpack PR.

silverwind commented:
@zeripath Seems like we are in quite dire situation. Fine with me then.

zeripath commented:
@silverwind Yeah it's a pretty poor situation. It would have likely been less annoying if we'd brought in semantic and it's less and JavaScript - and then Minified everything together as we could keep our own patches to semantic. 

It's possible that switching over to fomantic is not actually that difficult - just difficult for me when I'm trying to do other things. If you have any free cycles it's probably worth a go - then at least we'd have a working upstream.

Jookia commented:
I'm happy to try and push these changes to Fomantic if Gitea switches to it.

zeripath commented:
@Jookia I'm sure they would appreciate them even if we don't switch.

Jookia commented:
I'm sure they would too, but I don't have a way to test patches for Fomantic without a project that uses Fomantic

silverwind commented:
What is submodule `/public/vendor/plugins/semantic/Semantic-UI` for? We never submoduled JS dependencies and it does not seem to be referenced anywhere. Can it be removed?

Jookia commented:
Good catch, that was a mistake as I had it checked out there on my dev machine. Removed.

Jookia commented:
So this is short another approval. I asked @guillep2k but frontend isn't his speciality, so it looks like it might be @silverwind or some other maintainer that will have to do it.

zeripath commented:
To be clearer to help people discover the differences between semantic 2.3.1's dropdown and @Jookia s changes. Here's a unified diff:

```diff
--- semanti.dropdown.js.original	2018-03-19 04:04:27.000000000 +0000
+++ semantic.dropdown.js	2019-11-03 20:24:41.929563833 +0000
@@ -8,6 +8,13 @@
  *
  */
 
+/*
+ * Copyright 2019 The Gitea Authors
+ * Released under the MIT license
+ * http://opensource.org/licenses/MIT
+ * This version has been modified by Gitea to improve accessibility.
+ */
+
 ;(function ($, window, document, undefined) {
 
 'use strict';
@@ -33,6 +40,7 @@
     query          = arguments[0],
     methodInvoked  = (typeof query == 'string'),
     queryArguments = [].slice.call(arguments, 1),
+    lastAriaID = 1,
     returnedValue
   ;
 
@@ -114,6 +122,8 @@
 
             module.observeChanges();
             module.instantiate();
+
+            module.aria.setup();
           }
 
         },
@@ -296,6 +306,86 @@
           }
         },
 
+        aria: {
+          setup: function() {
+            var role = module.aria.guessRole();
+            if( role !== 'menu' ) {
+              return;
+            }
+            $module.attr('aria-busy', 'true');
+            $module.attr('role', 'menu');
+            $module.attr('aria-haspopup', 'menu');
+            $module.attr('aria-expanded', 'false');
+            $menu.find('.divider').attr('role', 'separator');
+            $item.attr('role', 'menuitem');
+            $item.each(function (index, item) {
+              if( !item.id ) {
+                item.id = module.aria.nextID('menuitem');
+              }
+            });
+            $text = $module
+              .find('> .text')
+              .eq(0)
+            ;
+            if( $module.data('content') ) {
+              $text.attr('aria-hidden');
+              $module.attr('aria-label', $module.data('content'));
+            }
+            else {
+              $text.attr('id', module.aria.nextID('menutext'));
+              $module.attr('aria-labelledby', $text.attr('id'));
+            }
+            $module.attr('aria-busy', 'false');
+          },
+          nextID: function(prefix) {
+            var nextID;
+            do {
+              nextID = prefix + '_' + lastAriaID++;
+            } while( document.getElementById(nextID) );
+            return nextID;
+          },
+          setExpanded: function(expanded) {
+            if( $module.attr('aria-haspopup') ) {
+              $module.attr('aria-expanded', expanded);
+            }
+          },
+          refreshDescendant: function() {
+            if( $module.attr('aria-haspopup') !== 'menu' ) {
+              return;
+            }
+            var
+              $currentlySelected = $item.not(selector.unselectable).filter('.' + className.selected).eq(0),
+              $activeItem        = $menu.children('.' + className.active).eq(0),
+              $selectedItem      = ($currentlySelected.length > 0)
+                ? $currentlySelected
+                : $activeItem
+            ;
+            if( $selectedItem ) {
+              $module.attr('aria-activedescendant', $selectedItem.attr('id'));
+            }
+            else {
+              module.aria.removeDescendant();
+            }
+          },
+          removeDescendant: function() {
+            if( $module.attr('aria-haspopup') == 'menu' ) {
+              $module.removeAttr('aria-activedescendant');
+            }
+          },
+          guessRole: function() {
+            var
+              isIcon = $module.hasClass('icon'),
+              hasSearch = module.has.search(),
+              hasInput = ($input.length > 0),
+              isMultiple = module.is.multiple()
+            ;
+            if ( !isIcon && !hasSearch && !hasInput && !isMultiple ) {
+              return 'menu';
+            }
+            return 'unknown';
+          }
+        },
+
         setup: {
           api: function() {
             var
@@ -335,6 +425,7 @@
             if(settings.allowTab) {
               module.set.tabbable();
             }
+            $item.attr('tabindex', '-1');
           },
           select: function() {
             var
@@ -477,6 +568,8 @@
               return true;
             }
             if(settings.onShow.call(element) !== false) {
+              module.aria.setExpanded(true);
+              module.aria.refreshDescendant();
               module.animate.show(function() {
                 if( module.can.click() ) {
                   module.bind.intent();
@@ -499,6 +592,8 @@
           if( module.is.active() && !module.is.animatingOutward() ) {
             module.debug('Hiding dropdown');
             if(settings.onHide.call(element) !== false) {
+              module.aria.setExpanded(false);
+              module.aria.removeDescendant();
               module.animate.hide(function() {
                 module.remove.visible();
                 callback.call(element);
@@ -902,7 +997,7 @@
           ;
           if(hasSelected && !module.is.multiple()) {
             module.debug('Forcing partial selection to selected item', $selectedItem);
-            module.event.item.click.call($selectedItem, {}, true);
+            $selectedItem[0].click();
             return;
           }
           else {
@@ -1363,7 +1458,7 @@
               // allow selection with menu closed
               if(isAdditionWithoutMenu) {
                 module.verbose('Selecting item from keyboard shortcut', $selectedItem);
-                module.event.item.click.call($selectedItem, event);
+                $selectedItem[0].click();
                 if(module.is.searchSelection()) {
                   module.remove.searchTerm();
                 }
@@ -1380,7 +1475,7 @@
                   }
                   else if(selectedIsSelectable) {
                     module.verbose('Selecting item from keyboard shortcut', $selectedItem);
-                    module.event.item.click.call($selectedItem, event);
+                    $selectedItem[0].click();
                     if(module.is.searchSelection()) {
                       module.remove.searchTerm();
                     }
@@ -1405,6 +1500,7 @@
                         .closest(selector.item)
                           .addClass(className.selected)
                       ;
+                      module.aria.refreshDescendant();
                       event.preventDefault();
                     }
                   }
@@ -1421,6 +1517,7 @@
                         .find(selector.item).eq(0)
                           .addClass(className.selected)
                       ;
+                      module.aria.refreshDescendant();
                       event.preventDefault();
                     }
                   }
@@ -1445,6 +1542,7 @@
                     $nextItem
                       .addClass(className.selected)
                     ;
+                    module.aria.refreshDescendant();
                     module.set.scrollPosition($nextItem);
                     if(settings.selectOnKeydown && module.is.single()) {
                       module.set.selectedItem($nextItem);
@@ -1472,6 +1570,7 @@
                     $nextItem
                       .addClass(className.selected)
                     ;
+                    module.aria.refreshDescendant();
                     module.set.scrollPosition($nextItem);
                     if(settings.selectOnKeydown && module.is.single()) {
                       module.set.selectedItem($nextItem);
@@ -2399,6 +2498,7 @@
               module.set.scrollPosition($nextValue);
               $selectedItem.removeClass(className.selected);
               $nextValue.addClass(className.selected);
+              module.aria.refreshDescendant();
               if(settings.selectOnKeydown && module.is.single()) {
                 module.set.selectedItem($nextValue);
               }
```

Jookia commented:
Doesn't look like this is going to get reviewed, so I'm closing this PR and keeping these changes local.

techknowlogick commented:
re-opening as I'm reviewing it right now (I've been travelling for the past while so I haven't been able to get to bigger reviews)

Jookia commented:
Oh, thanks!

techknowlogick commented:
@Jookia thank you so much for your PR, we really appreciate it.

Jookia commented:
I'm glad to hear that. Thanks everyone here. Hopefully I'll find time to do some more accessibility PRs. :)

silverwind commented:
@Jookia did you file the PR to [dropdown.js](https://github.com/fomantic/Fomantic-UI/blob/develop/src/definitions/modules/dropdown.js) yet? I could have a look but would prefer if you as the author do it.

zeripath commented:
@silverwind you could apply the patch in https://github.com/go-gitea/gitea/pull/8638#issuecomment-549175290

Jookia commented:
@silverwind : I have stopped all work on accessibility in Gitea, and consider that dropdown.js patch unmaintained and probably done the wrong way

zeripath commented:
@Jookia Er ... It is maintained. It is in our repo and we will maintain it.

It might be incorrect but there's plenty of code in Gitea that's incorrect.

If you're not interested in pushing it up - one of us will do it.

I'm not sure what happened that has made you give up on us but I'm sorry that you have.

silverwind commented:
I don't think I understand enough of this patch to submit it to Fomantic myself. I'm thinking it might be better to just drop the customization.

We're loading a old version of dropdown.js on top of Fomantic's updated version. I'm pretty certain this will eventually break it in one of the next Fomanic updates and it already impact page performance.

zeripath commented:
Fine I'll do it myself.
